### PR TITLE
[LLM] langchain integrations and examples

### PR DESCRIPTION
**Pull request by shane-huang**
## Description

Add langchain integrations for BigDL-LLM
- LLM integration
  -  [x] add new params to API to accommodate more model families 
   - [x] enable llamacpp path
   - [x] enable gptneox path
   - [x] enable bloom path
- Embedding integration
  -  [x] add new params to API to accommodate more model families 
   - [x] enable llamacpp path
   - [x] enable gptneox path
   - [x] enable bloom path
- usecases.exaample:
    - [x] added README 
    - [x] llamacpp stream chat example run through (gpt4all-7b, vicuna-13b, redpajama-7b)
    - [x] docqa stream example run through (gpt4all-7b, vicuna-13b, redpajama-7b)
    - [ ] tune the prompt template of redpajama (need tokenizer fix)
    - [ ] bloom/pheonix example enabling (need to change the bloom python API to allow for lora_path, etc.)
- [x] added simple example using llama-cpp-python style API (example/simple.py)

### 1. Why the change?
We need this integration so that other people can use langchain to develop all sorts of applications on BigDL-LLM supported models. 

### 2. User API changes

No API change to BigDL-LLM. Just langchain users can now use BigDL-LLM as follows: 

#### llm
```python
from bigdl.llm.langchain.llms import BigdlLLM
from langchain import PromptTemplate, LLMChain
from langchain.callbacks.manager import CallbackManager
from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler


prompt = PromptTemplate(template="""{question}""", input_variables=["question"])
llm = BigdlLLM(model_path="model/ggml/gpt4all-model-q4_0.bin")
callback_manager = CallbackManager([StreamingStdOutCallbackHandler()])

llm_chain = LLMChain(prompt=prompt, llm=llm, callback_manager=callback_manager)

question = "What is AI?"
llm_chain.run(question)
```
#### Embeddings
```python
from bigdl.llm.langchain.embeddings import BigdlLLMEmbeddings

bigdl_embeddings = BigdlLLMEmbeddings(model_path="model/ggml/gpt4all-model-q4_0.bin")

text = "This is a test document."
query_result = bigdl_embeddings.embed_query(text)
doc_result = bigdl_embeddings.embed_documents([text])
```
### 3. Summary of the change 

new module added under `bigdl/llm/langchain/llms`

### 4. How to test?
- [x] Unit test
- [x] Application test

**NOTE: langchain needs to raise errors in validate_environment implemenation but this won't pass our stylecheck. So disable codestyle check for langchain integrations for now. **

### 5. New dependencies

- [x] New Python dependencies
       - langchain

### CODE REVIEW

File: python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+from typing import Any, Dict, Generator, List, Optional
+
+from pydantic import Field, root_validator
+
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+
+
+
+class BigDLLLM(LLM):
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py:57]: Maybe `BigdlLLM`?

File: python/llm/example/simple.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from bigdl.llm.ggml.model.llama import Llama
+
+model = Llama("model/ggml/gpt4all-model-q4_0.bin")
```
Comment (jason-dai) [inline python/llm/example/simple.py:3]: shall we use a different filename postfix than llama.cpp, as their formats are not compatible? @TheaperDeng @MeouSker77

File: python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+# limitations under the License.
+#
+
+# This would makes sure Python is aware there is more than one sub-package within bigdl,
+# physically located elsewhere.
+# Otherwise there would be module not found error in non-pip's setting as Python would
+# only search the first bigdl package and end up finding only one sub-package.
+
+# This file is adapted from 
+# https://github.com/hwchase17/langchain/blob/master/langchain/llms/BigDLLLM.py
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py:23]: is this correct?

**Comment by shane-huang** [comment]
> Does it support other models than LLaMA?

- GPTNEOX (redpajama) is supported and run through both stream chat and docqa. The responses does not look very good, so I may need to tune the prompt template for redpajama in the examples. 

- Bloom API support is not ready yet. I will enable bloom in langchain once it is ready.

File: python/llm/example/llama_simple.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# This would makes sure Python is aware there is more than one sub-package within bigdl,
+# physically located elsewhere.
+# Otherwise there would be module not found error in non-pip's setting as Python would
+# only search the first bigdl package and end up finding only one sub-package.
+
+from bigdl.llm.ggml.model.llama import Llama
```
Comment (jason-dai) [inline python/llm/example/llama_simple.py:22]: Change to
```python
from bigdl.llm.models import LLaMA
```

File: python/llm/example/redpajama_simple.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+from bigdl.llm.ggml.model import Gptneox
```
Comment (jason-dai) [inline python/llm/example/redpajama_simple.py:1]: Change to
```python
from bigdl.llm.models import GPTNeoX
```

File: python/llm/example/langchain/docqa.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+# create embeddings and store into vectordb
+embeddings = BigdlLLMEmbeddings(model_path=MODEL_PATH, backbone=BACKBONE, n_ctx=2048)
+docsearch = Chroma.from_texts(texts, embeddings, metadatas=[{"source": str(i)} for i in range(len(texts))]).as_retriever()
+
+#get relavant texts
+docs = docsearch.get_relevant_documents(query)
+
+    
+bigdl_llm = BigdlLLM(
+    model_path=MODEL_PATH, backbone=BACKBONE, n_ctx=2048, callback_manager=callback_manager
```
Comment (jason-dai) [inline python/llm/example/langchain/docqa.py:60]: use `model_family` instead of `backbone`

File: python/llm/example/redpajama_simple.py (hunk 5 above)
Comment (shane-huang) [reply python/llm/example/redpajama_simple.py:22]: - This new API is not ready yet. Once its ready I'll change it in langchain integrations instead. 
- These simple examples (e.g. `llama_simple` and `redpajama_simple` ) are just for validation purposes (langchain users won't use these llm APIs directly). I plan to make them into UTs once CI/CD is ready.  

@TheaperDeng please inform people who's working on examples to also use the new interface `bigdl.llm.models`.

File: python/llm/example/llama_simple.py (hunk 4 above)
Comment (shane-huang) [reply python/llm/example/llama_simple.py:22]: - This new API is not ready yet. Once its ready I'll change it in langchain integrations instead.
- These simple examples (e.g. llama_simple and redpajama_simple ) are just for validation purposes (langchain users won't use these llm APIs directly). I plan to make them into UTs once CI/CD is ready.

@TheaperDeng please inform people who's working on examples to also use the new interface bigdl.llm.models.

Comment (TheaperDeng) [reply python/llm/example/llama_simple.py:22]: @plusbang is working on the namespace change (`from bigdl.llm.models import LLaMA`).
CICD has now been enabled on Linux, we will work on windows CICD next week.

File: python/llm/example/langchain/README.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+## Run the examples
+
+### 1. Streaming Chat
+
+```bash
+python ./streamchat.py -m MODEL_PATH -x MODEL_FAMILY -t THREAD_NUM -q "What is AI?"
+```
+arguments info:
+- `-m MODEL_PATH`: path to the converted model
+- `-x MODEL_FAMILY`: the model family of the model specified in `-m`, available options are `llama`, `gptneox`,`bloom`,
```
Comment (jason-dai) [inline python/llm/example/langchain/README.md:32]: no bloom for now?

Comment (shane-huang) [reply python/llm/example/langchain/README.md:32]: will fix. Should we add some langchain integration usage in README like in this PR description? Or examples are enough?

File: python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+            llm = BigdlLLM(model_path="/path/to/llama/model")
+    """
+
+
+    model_family: str = "llama"
+    """the model family: currently supports llama, gptneox, and bloom."""
+
+
+    family_info = {
+        'llama': {'module': "bigdl.llm.models" , 'class': "Llama"},
```
Comment (shane-huang) [inline python/llm/src/bigdl/llm/langchain/llms/bigdlllm.py:73]: @jason-dai changed to new namespace bigdl.llm.models as you required.

**Pull request merged**